### PR TITLE
feat(timepicker): add timezones

### PR DESCRIPTION
**Pull request by skhamvon**
- Manage timezones for timepicker with osds-select
- Add controller to handle behaviour of the component
- Add more tests for more coverage

### CODE REVIEW

File: packages/components/timepicker/src/components/osds-timepicker/core/controller.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+  constructor(private readonly component: OsdsTimepicker) {
+  }
+
+  checkSeconds(withSeconds: boolean) {
+    if (withSeconds === false && (this.component.input as HTMLInputElement)?.value.match(/:/g)?.length === 2) {
+      const inputValue = (this.component.input as HTMLInputElement).value.split(':');
+      (this.component.input as HTMLInputElement).value = `${inputValue[0]}:${inputValue[1]}`
+    } else if (withSeconds === true && (this.component.input as HTMLInputElement)?.value.match(/:/g)?.length === 1) {
+      (this.component.input as HTMLInputElement).value = `${(this.component.input as HTMLInputElement).value}:00`
```
Comment (dpellier) [inline packages/components/timepicker/src/components/osds-timepicker/core/controller.ts:15]: TBD: I think we should stop this back and forth between component and controller:
```
(component)checkSeconds =>
  => call (controller) checkSeconds
    => use (component)input
```
Plus in this case we set the value of the component element, in the controller, in a check function.


Would be more clear (and easier to unit test) to have something like:
```
(component)
  @Watch('withSeconds')
  @Watch('value')
  onSecondsChange(withSeconds: boolean) {
    const value = this.controller.formatValue(this.input.value, withSeconds);

    if (value) {
      this.input.value
    }
  }

(controller)
formatValue(value: string, withSeconds: boolean): string {
  if ... {
     ...
     return ...
  } else if ... {
     ...
     return ...
  }
  return '';
}
```

File: packages/components/timepicker/src/components/osds-timepicker/core/controller.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+  handleCurrentTimezone () {
+    if (!this.component.currentTimezone) {
+      const browserTimezone = new Date().getTimezoneOffset() / 60 * -1;
+      const parsedTimezone = browserTimezone >= 0 ? `+${browserTimezone}` : browserTimezone.toString();
+      this.component.currentTimezone = ODS_TIMEZONES.find((timezone) => (timezone as string).indexOf(parsedTimezone) > 1)
+    }
+  }
+
+  private getTimezonesList(): readonly (ODS_TIMEZONE | string)[] {
```
Comment (dpellier) [inline packages/components/timepicker/src/components/osds-timepicker/core/controller.ts:32]: `string` ?

File: packages/components/timepicker/src/components/osds-timepicker/public-api.ts [hunk 3]
```diff
@@ -1,2 +1,3 @@
 export type { OdsTimepickerAttribute } from './interfaces/attributes';
+export type { ODS_TIMEZONE, ODS_TIMEZONES } from './constants/timezones';
```
Comment (dpellier) [inline packages/components/timepicker/src/components/osds-timepicker/public-api.ts:2]: no type here, those are actual usable values

**Review by dpellier — CHANGES_REQUESTED** [review]

File: packages/components/timepicker/src/components/osds-timepicker/core/controller.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  checkSeconds(withSeconds: boolean) {
+    if (withSeconds === false && (this.component.input as HTMLInputElement)?.value.match(/:/g)?.length === 2) {
+      const inputValue = (this.component.input as HTMLInputElement).value.split(':');
+      (this.component.input as HTMLInputElement).value = `${inputValue[0]}:${inputValue[1]}`
+    } else if (withSeconds === true && (this.component.input as HTMLInputElement)?.value.match(/:/g)?.length === 1) {
+      (this.component.input as HTMLInputElement).value = `${(this.component.input as HTMLInputElement).value}:00`
+    }
+  }
+
+  handleTimezones () {
```
Comment (aesteves60) [inline packages/components/timepicker/src/components/osds-timepicker/core/controller.ts:19]:
```suggestion
  handleTimezones (): void {
```

File: packages/components/timepicker/src/components/osds-timepicker/core/controller.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    const check = this.getTimezonesList();
+    this.component.timezonesList = [...check] as ODS_TIMEZONE[] ;
+  }
+
+  handleCurrentTimezone () {
```
Comment (aesteves60) [inline packages/components/timepicker/src/components/osds-timepicker/core/controller.ts:24]:
```suggestion
  handleCurrentTimezone (): void {
```

File: packages/components/timepicker/src/components/osds-timepicker/osds-timepicker.tsx [hunk 6]
```diff
@@ -10,11 +13,16 @@ import { DEFAULT_ATTRIBUTE } from './constants/default-attributes';
 })
 export class OsdsTimepicker implements OdsTimepickerAttribute {
   @Element() el!: HTMLElement;
-  inputEl?: HTMLInputElement;
+  input: (OsdsInput & HTMLElement) | HTMLInputElement | null = null;
```
Comment (aesteves60) [inline packages/components/timepicker/src/components/osds-timepicker/osds-timepicker.tsx:16]: I think the type is just OsdsInput | null

File: packages/components/timepicker/src/components/osds-timepicker/core/controller.spec.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        const valueWithSeconds = "11:11:11";
+        const valueWithoutSeconds = "11:11";
+        const input = document.createElement('osds-input');
+
+        setup({ input });
+
+        component.input.value = valueWithSeconds;
+        component.withSeconds = true;
+        controller.checkSeconds(component.withSeconds)
+        expect(component.input.value).toEqual(valueWithSeconds);
```
[the remark this answers is not in the document]
Comment (skhamvon) [reply packages/components/timepicker/src/components/osds-timepicker/core/controller.spec.ts]: changed

File: packages/components/timepicker/src/components/osds-timepicker/core/controller.spec.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+        component.input.value = valueWithoutSeconds;
+        component.withSeconds = false;
+        controller.checkSeconds(component.withSeconds)
+        expect(component.input.value).toEqual(valueWithoutSeconds);
```
[the remark this answers is not in the document]
Comment (skhamvon) [reply packages/components/timepicker/src/components/osds-timepicker/core/controller.spec.ts]: changed

File: packages/components/timepicker/src/components/osds-timepicker/core/controller.ts (hunk 2 above)
Comment (skhamvon) [reply packages/components/timepicker/src/components/osds-timepicker/core/controller.ts]: fixed

File: packages/components/timepicker/src/components/osds-timepicker/public-api.ts (hunk 3 above)
Comment (skhamvon) [reply packages/components/timepicker/src/components/osds-timepicker/public-api.ts]: removed type

File: packages/components/timepicker/src/components/osds-timepicker/core/controller.ts (hunk 4 above)
Comment (skhamvon) [reply packages/components/timepicker/src/components/osds-timepicker/core/controller.ts]: added

File: packages/components/timepicker/src/components/osds-timepicker/core/controller.ts (hunk 5 above)
Comment (skhamvon) [reply packages/components/timepicker/src/components/osds-timepicker/core/controller.ts]: added

File: packages/components/timepicker/src/components/osds-timepicker/osds-timepicker.tsx (hunk 6 above)
Comment (skhamvon) [reply packages/components/timepicker/src/components/osds-timepicker/osds-timepicker.tsx]: HTMLInputElement is still necessary to access to value

File: packages/components/timepicker/src/components/osds-timepicker/core/controller.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+class OdsTimepickerController {
+
+  constructor(private readonly component: OsdsTimepicker) {
+  }
+
+  formatValue(value: string, withSeconds: boolean): string {
+    if (withSeconds === false && value.match(/:/g)?.length === 2) {
+      const inputValue = value.split(':');
+      return `${ inputValue[0] }:${ inputValue[1] }`
+    } else if (withSeconds === true && value.match(/:/g)?.length === 1) {
```
Comment (aesteves60) [inline packages/components/timepicker/src/components/osds-timepicker/core/controller.ts:14]: no else need

**Review by aesteves60 — CHANGES_REQUESTED** [review]

**Review by dpellier — APPROVED** [review]

**Review by aesteves60 — APPROVED** [review]

**Review by astagnol — APPROVED** [review]

**Pull request merged**